feat: Langfuse integration, system-wide tracing configuration (LangSmith&Langfuse)

maciejmajek opened the pull request:
## Purpose

Tracing agents is crucial for debugging, transparency, safety and more. At the moment, the code only works with LangSmith tracing (a paid solution). RAI has to provide an alternative, like [LangFuse](https://github.com/langfuse/langfuse).

## Proposed Changes

What does this PR add, remove or fix?

## Issues

- Links to relevant issues

## Testing

Setup langfuse: https://langfuse.com/docs/deployment/local
Go to http://localhost:3000, create account, organization and project.
configure langfuse tracing in config.toml
```toml
[tracing.langfuse]
use_langfuse = true
host = "http://localhost:3000"
```
Add the config to the main agent.invoke
```python
from rai.utils.model_initialization import get_llm_model, get_tracing_callbacks
...
self.agent.invoke(
            {"messages": self.history}, config={"callbacks": get_tracing_callbacks()}
        )
```
<img width="360" alt="image" src="https://github.com/user-attachments/assets/1852c314-6456-4c29-86f8-2737a5c8da66">

<img width="360" alt="image" src="https://github.com/user-attachments/assets/278133bf-a5b4-4d61-ad4a-e7954c05b0fe">

maciejmajek committed:
feat: add tracing configuration in config.toml
> implement get_tracing_callbacks for system wide use cases

fix: proper config parsing

boczekbartek commented:
@maciejmajek Such a great idea to have an open-source alternative for paid services! 

Please document how to use Langfuse (and LangSmith if possible). I think new users might not know that it's possible. 

How about adding it to:
- `tracing.md` in https://github.com/RobotecAI/rai/tree/development/docs 
or  
- https://github.com/RobotecAI/rai/blob/development/docs/vendors.md
?

maciejmajek committed:
docs: add tracing documentation

Apply suggestions from code review

Co-authored-by: Adam Dąbrowski <adam.dabrowski@robotec.ai>

docs: rewrite tracing.md (prioritize open source solution), fix misleading info

maciejmajek commented:
Multimodal tracing works in langsmith, langfuse has limited capability (works with links, but not with base64 images that we use yet)
<img width="982" alt="image" src="https://github.com/user-attachments/assets/21a7f4b4-8d0a-4a30-b183-33b0dc7218f7">

adamdbrw commented on docs/tracing.md:
```diff
@@ -0,0 +1,55 @@
+The `project` field under the `[tracing]` section sets the name for your tracing project. This name will be used to identify your project in the tracing tools.
+
+> [!NOTE]  
+> Only LangSmith project name is currently used. Langfuse will upload traces to the default project.
```
```suggestion
> Project name is currently only used by LangSmith. Langfuse will upload traces to the default project.
```

adamdbrw commented on docs/tracing.md:
```diff
@@ -0,0 +1,55 @@
+
+## Configuration
+
+Tracing configuration is managed through the `config.toml` file. The relevant section for tracing is as follows:
```
```suggestion
Tracing configuration is managed through the `config.toml` file. The relevant parameters for tracing are:
```

adamdbrw commented on docs/tracing.md:
```diff
@@ -0,0 +1,55 @@
+
+When enabled, LangSmith tracing will automatically be set up with the following environment variables:
+
+### Langfuse
```
I would always start with the open source option when reading from the top, and make the distinction extra clear, i.e.:

### Langfuse (open-source)
### LangSmith (closed-source, paid)

adamdbrw commented on docs/tracing.md:
```diff
@@ -0,0 +1,55 @@
+# Tracing Configuration
+
+RAI supports tracing capabilities to help monitor and analyze the performance of your AI models. This document outlines how to configure tracing for your RAI project.
```
```suggestion
RAI supports tracing capabilities to help monitor and analyze the performance of your AI models, at a minor performance cost. By default, tracing is off. This document outlines how to configure tracing for your RAI project.
```

adamdbrw commented on docs/tracing.md:
```diff
@@ -0,0 +1,55 @@
+If you encounter issues with tracing:
+
+1. Ensure all required environment variables are set correctly.
+2. Check that the `use_langsmith` and `use_langfuse` flags are set to `true` in `config.toml` if you intend to use these services.
```
```suggestion
2. Check whether tracing is on by checking whether `use_langsmith` or `use_langfuse` flag is set to `true` in `config.toml`.
```

adamdbrw commented on docs/developer_guide.md:
```diff
@@ -94,6 +94,7 @@ agent.invoke({"messages": ["Please pick up an object and scan it."]})
 
 Additional resources:
 
+- [Tracing](tracing.md) configuration.
```
```suggestion
- [Tracing](tracing.md) configuration for AI models.
```

maciejmajek replied:
```suggestion
- [Tracing](tracing.md) configuration for genAI models and agents.
```

adamdbrw reviewed (approved):
Some suggestions, rather minor

adamdbrw commented:
Important step for full open-source tool set
